### ISSUE TITLE
Versioning the specification

### ISSUE THREAD

**Issue by kvark** [KhronosGroup/Vulkan-Portability#12]
We need to figure out how the specification is going to be developed now that we ratified version 1 of the provisional spec. At some point, we'll need to release a non-provisional v1, backed by the CTS coverage of some implementations. But then we'll still keep discovering new limitations, potentially, especially in D3D12, which we don't have as clear of an idea about, at this point. The existing structures are set in stone after v1, and we'll have to add new structures.

Are we going to make it a version 2 of the same specification, with new structures in it? Is it even valid to add new types/structures to later versions of a spec, as long as backwards compatibility is preserved? cc @dgkoch

**Comment by danginsburg** [comment]
> Is it even valid to add new types/structures to later versions of a spec, as long as backwards compatibility is preserved?

_Caveat: I could totally be wrong here, but this is my understanding_ 

I am not aware of any precedent for doing this.  The VkExtensionProperties.specVersion can be queried in the API but I don't believe there is any way for an application to tell the driver which version it wants.  I think the only mechanism for new type/structures after provisional ratification is a new extension.

**Comment by kvark** [comment]
@danginsburg I believe you are correct here (and I might have said on one of the calls that you can select a version, that was a wrong statement). However, what's the danger in a situation where the user only knows about V1 but the driver exposes V2 with a few extra structures?

**Comment by danginsburg** [comment]
@kvark One of the principles of the Vulkan extension mechanism is that a driver should always be informed of all extensions/features the application intends to use.  That's why the ppEnabledExtensions and features are passed to vkCreateDevice.  In OpenGL the driver wouldn't know which extensions were in use which would mean it might pay a cost to have a feature when an application hasn't used it.  So at a minimum it's in violation of that philosophy. 

I think there are practical considerations too.  For example, what does validation check?  Does it check all things associated with v2 always?  That means an app that previously passed validation with our extension would become invalid if a new VU check was added for v2 and it violated it, unless there was also a rule that validation only checked v2 if its related structures existed in the pNext chain, but that language would have to be specific to our spec and probably opens a can of worms in terms of allowing any extension to do that.  And what if it's not something in the pNext change but a new enum?  Now it really has no way of knowing.

My guess is it won't fly.

**Comment by dgkoch** [comment]
@danginsburg is mostly correct, with one caveat. 

There is an opportunity to change an extension's functionality when it moves from provisional to final (we did this for the ray tracing extensions). Not to say that doesn't come with it's own pain, but that's the point of provisional releases. 
Once an extension has been released as final there is no way to keep adding (or in this case, adding removals!?) to it. You'll need to use another extension.

**Comment by kvark** [comment]
Thank you for responses! If I understand correctly, the main problem here is backwards compatibility. So if an app is written against V1 of the spec, and the implementation has V2, we can't suddenly make it fail to validate.

Is there a way to get that behavior while still adding new structures then? For example, can we specify this as "if the user requests structure Xxxx, then the limits returned are validated, otherwise they are ignored"? Alternatively, a boolean field may be added to those structure with something like "enableValidation" semantics, and it's zero by default.

**Comment by dgkoch** [comment]
> Is there a way to get that behavior while still adding new structures then?

No. There is no way to add new structures without a new extension. When I originally advocated for adding the specVersion field, I thought it might be possible, but we've later decided that shouldn't be done. 

For spec references see StyleGuide section 3.12 ["Required Extension Tokens"](https://www.khronos.org/registry/vulkan/specs/1.2/styleguide.html#_required_extension_tokens)

> "The revision value indicates a patch version of the extension specification, and differences in this version number maintain full compatibility, as defined in the Compatibility Guarantees section of the Vulkan API Specification."

and 39.6.1 ["Core Versions: Patch versions"](https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#extendingvulkan-compatibility-coreversions)
> "Specifications with different patch versions but the same major and minor version are fully compatible with each other - such that a valid application written against one will work with an implementation of another."

> "Note If a patch version includes a bug fix or clarification that could have a significant impact on developer expectations, these will be highlighted in the change log. Generally the Vulkan Working Group tries to avoid these kinds of changes, instead fixing them in either an extension or core version."

**Comment by kvark** [comment]
Ok, thanks @dgkoch ! Looks like the spec is pretty strict about this, as you say:
> Any changes requiring the addition or removal of a type or command should be done by creating a new extension. The resulting extension should take care to include the appropriate dependency information on the original extension.

**Comment by krOoze** [comment]
What about making the feature set open? E.g. an array of {enum, bool} pairs, where it would be allowed to add feature enumerants.

Then app would be instructed in the spec to iterate through the features and if it encounters `VK_FALSE` in any enum that it does not know, then it **should** refuse to use that device.
